File: trunk/bx-src/libbx.pas/libbx_stream.pas
{
  Copyright (C) 2013 Melchiorre Caruso

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
}

{
  Contains:

  Modifyed:
}

unit libbx_Stream;

interface

const
  { Default buffer capacity }

  DEFAULT_BUFFER_CAPACITY = 4096;

type
  { I/O functions }

  PStreamRead  = function(Stream: Pointer; Data: PByte; Count: longint): longint;
  PStreamWrite = function(Stream: Pointer; Data: PByte; Count: longint): longint;

type
  { TReadStream struct/methods }

  PReadStream = ^TReadStream;

  TReadStream = packed record
    Stream: Pointer;
    StreamRead: PStreamRead;
    BufferSize: longint;
    BufferReaded: longint;
    Buffer: array[0.. DEFAULT_BUFFER_CAPACITY - 1] of byte;
  end;

  function  ReadStream_Create(aStream: Pointer; aStreamRead: PStreamRead): PReadStream;
  procedure ReadStream_Destroy    (Self: PReadStream);
  procedure ReadStream_ClearBuffer(Self: PReadStream);
  procedure ReadStream_FillBuffer (Self: PReadStream);
  function  ReadStream_Read       (Self: PReadStream): byte;

type
  { TWriteStream struct/methods }

  PWriteStream = ^TWriteStream;

  TWriteStream = packed record
    Stream: Pointer;
    StreamWrite: PStreamWrite;    
    BufferSize: longint;
    Buffer: array[0.. DEFAULT_BUFFER_CAPACITY - 1] of byte;
  end;

  function  WriteStream_Create(aStream: Pointer; aStreamWrite: PStreamWrite): PWriteStream;
  procedure WriteStream_Destroy    (Self: PWriteStream);
  procedure WriteStream_ClearBuffer(Self: PWriteStream);
  procedure WriteStream_FlushBuffer(Self: PWriteStream);
  procedure WriteStream_Write      (Self: PWriteStream; Data: byte);

implementation

{ TReadStream methods }

function ReadStream_Create(aStream: Pointer; aStreamRead: PStreamRead): PReadStream;
begin
  ReadStream_Create := GetMem(sizeof(TReadStream));
  ReadStream_Create^.Stream       := aStream;
  ReadStream_Create^.StreamRead   := aStreamRead;
  ReadStream_Create^.BufferSize   := 0;
  ReadStream_Create^.BufferReaded := 0;
end;

procedure ReadStream_Destroy(Self: PReadStream);
begin
  FreeMem(Self);
end;

procedure ReadStream_ClearBuffer(Self: PReadStream);
begin
  Self^.BufferReaded := 0;
  Self^.BufferSize   := 0;
end;

procedure ReadStream_FillBuffer(Self: PReadStream);
begin
  Self^.BufferSize   := Self^.StreamRead(Self^.Stream, @Self^.Buffer[0], DEFAULT_BUFFER_CAPACITY);
  Self^.BufferReaded := 0;
end;

function ReadStream_Read(Self: PReadStream): byte;
begin
  if Self^.BufferReaded < Self^.BufferSize then
  begin
    ReadStream_Read := Self^.Buffer[Self^.BufferReaded];
    Inc(Self^.BufferReaded);
  end else
  begin
    ReadStream_FillBuffer(Self);
    if Self^.BufferReaded < Self^.BufferSize then
    begin
      ReadStream_Read := Self^.Buffer[Self^.BufferReaded];
      Inc(Self^.BufferReaded);
    end;
  end;
end;

{ TWriteStream methods}

function WriteStream_Create(aStream: Pointer; aStreamWrite: PStreamWrite): PWriteStream;
begin
  WriteStream_Create := GetMem(sizeof(TWriteStream));
  WriteStream_Create^.Stream      := aStream;
  WriteStream_Create^.StreamWrite := aStreamWrite;
  WriteStream_Create^.BufferSize  := 0;
end;

procedure WriteStream_Destroy(Self: PWriteStream);
begin
  WriteStream_FlushBuffer(Self);
  FreeMem(Self);
end;

procedure WriteStream_ClearBuffer(Self: PWriteStream);
begin
  Self^.BufferSize := 0;
end;

procedure WriteStream_FlushBuffer(Self: PWriteStream);
begin
  if Self^.BufferSize <> 0 then
  begin
    Self^.StreamWrite(Self^.Stream, @Self^.Buffer[0], Self^.BufferSize);
    Self^.BufferSize := 0;
  end;
end;

procedure WriteStream_Write(Self: PWriteStream; Data: byte);
begin
  if Self^.BufferSize = DEFAULT_BUFFER_CAPACITY then
  begin
    WriteStream_FlushBuffer(Self);
  end;
  Self^.Buffer[Self^.BufferSize] := Data;
  Inc(Self^.BufferSize);
end;

end.

File: trunk/bx-src/libbx_stream.c
/*
  Copyright (c) 2012-2014 Melchiorre Caruso

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the free software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT any WARRANTY; WITHOUT even the implied WARRANTY of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License FOR more details.

  you should have received A copy of the GNU General Public License
  along with This program; if not, write to the free software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

/*
  Contains:

    Common buffered stream.

  Modifyed:

    v1.0.0 build 2202 - 2014.01.13 by Melchiorre Caruso.

*/

#include "libbx_stream.h"
#include <stdlib.h>

/* Default buffer capacity */

#define DEFAULT_BUFFER_CAPACITY 4096L

/* TReadStream struct/methods implemetation */

struct TReadStream {
         void *Stream;
    PStreamRead StreamRead;
       int32_t BufferSize;
       int32_t BufferReaded;
       uint8_t Buffer[DEFAULT_BUFFER_CAPACITY - 1];
};

PReadStream ReadStream_Create(void *aStream, PStreamRead aStreamRead)
{
  PReadStream Self   = malloc(sizeof(struct TReadStream));
  Self->Stream       = aStream;
  Self->StreamRead   = aStreamRead;

  Self->BufferReaded = 0;
  Self->BufferSize   = 0;
  return Self;
}

void ReadStream_Destroy(PReadStream Self)
{
  free(Self);
}

inline void ReadStream_ClearBuffer(PReadStream Self)
{
  Self->BufferReaded = 0;
  Self->BufferSize   = 0;
}

inline void ReadStream_FillBuffer(PReadStream Self)
{
  Self->BufferSize   = Self->StreamRead(Self->Stream, &(Self->Buffer[0]), DEFAULT_BUFFER_CAPACITY);
  Self->BufferReaded = 0;
}

inline uint8_t ReadStream_Read(PReadStream Self)
{
  uint8_t result = 0;
  if (Self->BufferReaded < Self->BufferSize)
  {
    result = Self->Buffer[Self->BufferReaded];
    Self->BufferReaded++;
  }
  else
  {
    ReadStream_FillBuffer(Self);
    if (Self->BufferReaded < Self->BufferSize)
    {
      result = Self->Buffer[Self->BufferReaded];
      Self->BufferReaded++;
    }
  }
  return result;
}

/* TWriteStream struct/methods implemetation */

struct TWriteStream {
         void *Stream;
  PStreamWrite StreamWrite;
         int32_t BufferSize;
         uint8_t Buffer[DEFAULT_BUFFER_CAPACITY - 1];
};

PWriteStream WriteStream_Create(void *aStream, PStreamWrite aStreamWrite)
{
  PWriteStream Self = malloc(sizeof(struct TWriteStream));
  Self->Stream      = aStream;
  Self->StreamWrite = aStreamWrite;
  Self->BufferSize  = 0;
  return Self;
}

void WriteStream_Destroy(PWriteStream Self)
{
  free(Self);
}

inline void WriteStream_ClearBuffer(PWriteStream Self)
{
  Self->BufferSize = 0;
}

inline void WriteStream_FlushBuffer(PWriteStream Self)
{
  if (Self->BufferSize != 0)
  {
    Self->StreamWrite(Self->Stream, &(Self->Buffer[0]), Self->BufferSize);
    Self->BufferSize = 0;
  }
}

inline void WriteStream_Write(PWriteStream Self, uint8_t Data)
{
  if (Self->BufferSize == DEFAULT_BUFFER_CAPACITY)
  {
    WriteStream_FlushBuffer(Self);
  }
  Self->Buffer[Self->BufferSize] = Data;
  Self->BufferSize++;
}

